fix button bug for 2.1-release

PPpro opened the pull request:
Re: https://github.com/cocos-creator/2d-tasks/issues/399

改动：
创建 State 枚举
添加 node  COLOR_CHANGED 事件

修复问题：
禁用 interactable， normalSprite 会设置成 diabledSprite (修复同时 reset 了之前的一个提交 https://github.com/cocos-creator/engine/pull/808)
按钮 color 模式下，设置透明度无效
优化编辑器修改 button  transition 的体验
- transition 从 color 改成其他模式，node 上的颜色应该再初始化（或者变灰度状态）
- transition 从 sprite 改成其他模式，spriteFrame 应该设置为 Normal 状态
- 灰度状态下 改成 color 模式，node 上的颜色改成 disbaled 状态的颜色
- 灰度状态下 改成 sprite 模式，spriteFrame 应该设置为 disabled 状态（如果存在）

pandamicro commented on cocos2d/core/CCNode.js:
```diff
@@ -944,6 +944,8 @@ var Node = cc.Class({
                         this._renderFlag |= RenderFlow.FLAG_COLOR;
                     }
+
+                    this.emit('color-changed', value);
```
如果要添加事件，遵循其他 NodeEvent 的判断和派发方式

PPpro replied:
这个我再规范化一下

pandamicro commented on cocos2d/core/components/CCButton.js:
```diff
@@ -235,6 +237,11 @@ let Button = cc.Class({
+            notify () {
+                if (this.transition === Transition.Color && this._getButtonState() === State.PRESSED) {
+                    this.target.opacity = this.pressedColor.a;
+                }
```
为什么 pressedColor 和 hoverColor 不需要 updateState ？

PPpro replied:
这个是针对编辑器的优化，编辑器里只能看到 normalColor 和 disabledColor
用户需要在编辑器里修改完 color 之后实时看到更新后的效果，所以需要 对 normalColor disabledColor 的修改进行 updateState，pressedColor 和 hoverColor 就不需要

PPpro replied:
好吧，这里 pressedColor 和 hoverColor 还是有必要修改
如果 transition duration 比较长的话， 在这段时间修改 pressedColor / hoverColor， 过度颜色会有问题

pandamicro commented on cocos2d/core/components/CCButton.js:
```diff
@@ -429,7 +444,14 @@ let Button = cc.Class({
+            this.node.on('color-changed', function (color) {
+                if (this.transition === Transition.COLOR) {
+                    let state = this._getButtonState();
+                    this[state + 'Color'] = color;
```
不要用字符串拼接，用 switch 的方式

pandamicro commented on cocos2d/core/components/CCButton.js:
```diff
@@ -642,6 +666,17 @@ let Button = cc.Class({
+            this._updateColorTransition(State.NORMAL);
+        }
+        else if (oldTransition === Transition.SPRITE) {
+            this._updateSpriteTransition(State.NORMAL);
```
这里应该用当前 state 吧？

PPpro replied:
这样做的原因是，当sprite 处于 disabledSprite 的情况下，切换到 比如 color 模式，这时候 sprite 应该改回 normalSprite
color transition 也是一样的处理方式

pandamicro replied:
哦哦，理解了

pandamicro reviewed (changes requested):
如果在 transition 过程中发生各个 state 下 color 或者 frame 的改变，会不会不生效

PPpro committed:
fix button bug for 2.1-release

create button state enum
fix setNormalSprite when change button interactable in editor
fix button can't set opacity in color transition
reset button to normal data when change transition
reset pr #808
add node COLOR_CHANGED event
